corehq/apps/userreports/pillow.py

```diff
@@ -208,6 +211,100 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+            return [docs_by_id.get(change.id) for change in good_changes]
+
+        self.bootstrap_if_needed()
+
+        # break up changes by domain
+        changes_by_domain = defaultdict(list)
+        changes_by_id = {change.id: change for change in changes}
+        for change in changes:
+            if change.metadata.domain and change.metadata.domain in self.table_adapters_by_domain:
+                # if no domain we won't save to any UCR table
+                changes_by_domain[change.metadata.domain].append(change)
+
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            rows_to_save_by_adapter = defaultdict(list)
+            adapters = list(self.table_adapters_by_domain[domain])
+            to_delete = {change for change in changes_chunk if change.deleted}
+            to_update = set(changes_chunk) - to_delete
+            docs = get_docs(to_update, domain)
+            async_configs_by_doc_id = defaultdict(list)
+
+            for doc in docs:
+                eval_context = EvaluationContext(doc)
+                for adapter in adapters:
+                    if adapter.config.filter(doc):
+                        if adapter.run_asynchronous:
+                            async_configs_by_doc_id[doc['_id']].append(adapter.config._id)
+                        else:
+                            try:
+                                rows_to_save_by_adapter[adapter].extend(adapter.get_all_values(doc, eval_context))
+                            except Exception as ex:
+                                notify_exception(None,
+                                    "{name} Error in getting UCR values for changes chunk {ids}: {ex}".format(
+                                        name=pillow_instance.get_name(), ids=[c.id for c in to_delete], ex=ex))
+                                failed_changes.add(changes_by_id[doc['_id']])
+                    elif adapter.config.deleted_filter(doc):
+                        to_delete.add(changes_by_id[doc['_id']])
+                eval_context.reset_iteration()
+
+            for adapter in adapters:
+                try:
+                    adapter.bulk_delete([c.id for c in to_delete])
+                except Exception as ex:
+                    notify_exception(None,
+                        "{name} Error in deleting changes chunk {ids}: {ex}".format(
+                            name=pillow_instance.get_name(), ids=[c.id for c in to_delete], ex=ex))
+                    failed_changes.union(to_delete)
+            for adapter, rows in six.iteritems(rows_to_save_by_adapter):
+                try:
+                    adapter.save_rows(rows)
+                except Exception:
+                    notify_exception(None,
+                        "{name} Error in saving changes chunk {ids}: {ex}".format(
+                            name=pillow_instance.get_name(), ids=[c.id for c in to_delete], ex=ex))
+                    failed_changes.union(to_update)
```

Comment:
this just marks the all as failed including the async ones. would be nice if we could make it more specific